modules/core/src/main/java/org/apache/ignite/internal/pagemem/wal/record/UnwrappedDataEntry.java

```diff
@@ -35,4 +41,86 @@
      * @return Value which was placed into cache. Or null for delete operation or for failure.
      */
     Object unwrappedValue();
+
+    /**
+     * Returns a string representation of the entry.
+     *
+     * @param entry          Object to get a string presentation for.
+     * @param superToString  String representation of parent.
+     * @param cacheObjValCtx Cache object value context. Context is used for unwrapping objects.
+     * @param <T>            Composite type: extends DataEntry implements UnwrappedDataEntry
+     * @return String presentation of the given object.
+     */
+    public static <T extends DataEntry & UnwrappedDataEntry> String toString(
+        T entry,
+        String superToString,
+        final CacheObjectValueContext cacheObjValCtx
+    ) {
+        final Object key = entry.unwrappedKey();
+
+        String keyStr;
+        if (key instanceof String) {
+            keyStr = (String)key;
+        }
+        else if (key instanceof BinaryObject) {
+            keyStr = key.toString();
+        }
+        else {
+            keyStr = (key != null) ? toStringRecursive(key.getClass(), key) : null;
+        }
+
+        if (keyStr == null || keyStr.isEmpty()) {
+            try {
+                keyStr = Base64.getEncoder().encodeToString(entry.key().valueBytes(cacheObjValCtx));
+            }
+            catch (IgniteCheckedException e) {
+                cacheObjValCtx.kernalContext().log(UnwrapDataEntry.class)
+                    .error("Unable to convert key [" + entry.key() + "]", e);
+            }
+        }
+
+        final Object value = entry.unwrappedValue();
+
+        String valueStr;
+        if (value instanceof String) {
+            valueStr = (String)value;
+        }
+        else
+        if (value instanceof BinaryObject) {
+            valueStr = value.toString();
+        }
+        else {
+            valueStr = (value != null) ? toStringRecursive(value.getClass(), value) : null;
+        }
+
+        if (valueStr == null || valueStr.isEmpty()) {
+            try {
+                valueStr = Base64.getEncoder().encodeToString(entry.value().valueBytes(cacheObjValCtx));
+            }
+            catch (IgniteCheckedException e) {
+                cacheObjValCtx.kernalContext().log(UnwrapDataEntry.class)
+                    .error("Unable to convert value [" + entry.value() + "]", e);
+            }
+        }
+
+        return entry.getClass().getSimpleName() + "[k = " + keyStr + ", v = ["
```

Comment:
I think we should add flag to be able to hide sensitive data in WAL reader output. For example, we can use cryptographic hash to hide data and see some representation of it to know is it unique or not, with high probability.

Reply:
No need? because if not set binary metadata object fields will not be displayed (parameter is implemented within different issue), only its binary representation

Reply:
Security policies don't allow us to receive WAL reader output even when there is only binary representation, because there is theoretical possibility to decode data. If we will have only hashes instead of data in any representation, there will be no need to deny us from receiving WAL reader output.

Reply:
Create https://ggsystems.atlassian.net/browse/GG-26945